src/nrtk_explorer/app/ui/image_list.py

```diff
@@ -153,26 +106,61 @@ def __init__(
             )
 
 
+@TrameApp()
 class ImageList(html.Div):
-    instances = []
-
-    @staticmethod
-    def reset_view_range():
-        for instance in ImageList.instances:
-            instance.visible_ids = set()
-            server.js_call(ref="image-list", method="resetVirtualScroll")
 
     def set_in_view_ids(self, ids):
         visible = set(ids)
         if self.visible_ids != visible:
             self.visible_ids = visible
             self.scroll_callback(self.visible_ids)
 
-    def __init__(self, on_scroll, on_hover, **kwargs):
+    def _set_image_list_ids(self, dataset_ids):
+        # create reactive variables so ImageDetection components have live Refs
+        for id in dataset_ids:
+            keys = get_image_state_keys(id)
+            for key in keys.values():
+                if not self.state.has(key):
+                    self.state[key] = None
+        self.state.image_list_ids = dataset_ids
+
+    @change("dataset_ids", "user_selected_ids")
+    def update_image_list_ids(self, **kwargs):
+        if len(self.state.user_selected_ids) > 0:
+            self._set_image_list_ids(self.state.user_selected_ids)
+        else:
+            self._set_image_list_ids(self.state.dataset_ids)
+
+    @change("image_list_ids")
+    def reset_view_range(self, **kwargs):
+        self.visible_ids = set()
+        self.server.js_call(ref="image-list", method="resetVirtualScroll")
+        if self.state.image_list_view_mode == "grid":
+            self.server.controller.get_visible_ids()
+
+    @change("image_list_view_mode")
+    def update_pagination(self, **kwargs):
+        old_pagination = self.state.pagination or {}
+        if self.state.image_list_view_mode == "grid":
+            self.state.pagination = {**old_pagination, "rowsPerPage": 12}
+            self.server.controller.get_visible_ids()
+        else:
+            self.state.pagination = {**old_pagination, "rowsPerPage": 0}  # show all rows
+
+    def __init__(self, server, on_scroll, on_hover, **kwargs):
         super().__init__(classes="full-height", **kwargs)
-        ImageList.instances.append(self)
+        self.server = server
+        self.state = server.state
+        self.ctrl = server.controller
+
```

Comment:
Because you inherit `html.Div`, you do not need to pass the server and set state/ctrl. [see](https://github.com/Kitware/trame-client/blob/master/trame_client/widgets/core.py#L335-L357)

Reply:
🕵️✅